nca/FWRules/ConnectivityGraph.py

```diff
@@ -133,26 +129,133 @@ def _creates_cliqued_graph(directed_edges):
+        bicliques_edges = set()
+
+        while directed_edges:
+            # find dst_sets:
+            all_sources = set(edge[0] for edge in directed_edges)
+            src_to_dst_set = {src: frozenset(e[1] for e in directed_edges if e[0] == src) for src in all_sources}
+            all_dst_set = frozenset(src_to_dst_set.values())
+
+            # find all bicliques:
+            all_bicliques = []
+            for dst_set in all_dst_set:
+                src_set = frozenset(src for src in all_sources if src_to_dst_set[src] >= dst_set)
+                all_bicliques.append((src_set, dst_set))
+            # find the best biclique:
+            bicliques_ranks = {(s, d): len(s) * len(d) - len(s) - len(d) for s, d in all_bicliques}
+            if max(bicliques_ranks.values()) < 1:
+                break
+            best_biclique = max(bicliques_ranks, key=bicliques_ranks.get)
+
+            # create the new biclique, and remove it from the graph:
+            directed_edges -= set(itertools.product(best_biclique[0], best_biclique[1]))
+            biclique_name = f'biclique_{conn_str}{len(bicliques_nodes)}'
+            biclique_node = (biclique_name, '')
+            bicliques_nodes.append(biclique_node)
+            bicliques_edges |= set((src, biclique_node) for src in best_biclique[0])
+            bicliques_edges |= set((biclique_node, dst) for dst in best_biclique[1])
+        return directed_edges, bicliques_edges, bicliques_nodes
+
+    @staticmethod
+    def _find_equal_groups(peers_edges):
+        """
+        find set of peers that have the same io edges and the same namespace
+        :param: peers_edges: dict peer -> [peer edges]
+        return: list
+        list: list of lists, each list is a group of peers
+        list: list of peers that are not in any group
+        """
+        groups_dict = defaultdict(list)
+        for peer, peer_edges in peers_edges.items():
+            groups_dict[(peer.namespace, peer_edges)].append(peer)
+        groups = [sorted(peer, key=str) for peer in groups_dict.values() if len(peer) > 1]
+        left_out = [peer[0] for peer in groups_dict.values() if len(peer) == 1]
+        return groups, left_out
+
+    def _get_equals_groups(self):
+        """
+        _get_equals_groups find in the graph sets of peers that has the same connections.
+        it find two kinds of groups:
+        1. set of peers that have the same input and the same output edges
+        2. set of peers that have the same input and the same output edges and also connected to each other
+
+        the algorithm:
+        1. for each peer, collect the list of input and output edges.
+        2. for each peer, add self pointing edge
+        3. find group of peers that share the same edges list (the group of kind (2))
+        4. remove the peers that are already grouped
+        5. remove the self pointing edge
+        6. find group of peers that share the same edges list (the group of kind (1))
```

Comment:
Wouldn't it make more sense to first build kind (1) groups, then add self edges, then build kind (2) groups?

Reply:
if a peer is in a group of kind (1), and in another peer of kind (2). I prefer to put in on the kind (2) group

Reply:
Good point. Then your implementation is correct.